pallene/assignment_conversion.lua

```diff
@@ -143,17 +148,48 @@ function Converter:apply_transformations()
             --    to ast.Var.Dot
             local typ = types.T.Record(self:type_name(decl.name), { "value" } , { value = decl._type } )
             self:add_box_type(decl.loc, typ)
-            decl._type = typ
 
             local init_exp_update = self.update_init_exp_of_decl[decl]
 
             if init_exp_update then
+                decl._type = typ
                 local old_ref = init_exp_update.ref
                 local update  = init_exp_update.update_fn
 
                 local new_node = ast.Exp.InitList(old_ref.loc, {{ name = "value", exp = old_ref }})
                 new_node._type = typ
                 update(new_node)
+
+            elseif self.lambda_of_param[decl] then
+                --- Function parameters that are captured are implementing by "proxy"-ing them.
+                --- Consider the following function that returns a closure:
+                --- ```
+                --- function m.foo(n: integer)
+                ---   -- capture n
+                --- end
+                --- ```
+                --- Since we cannot transform the declaration node of a function parameter,
+                --- we create a variable to represent the boxed parameter which can be captured.
+                --- ```
+                --- function m.foo(n: integer)
+                ---   local $n: $T = { value = n }
+                ---   -- capture and mutate $n
+                --- end
+                local param = ast.Exp.Var(decl.loc, ast.Var.Name(decl.loc, decl.name))
+                param.var._def = checker.Def.Variable(decl)
+                param.var._type = decl._type
+                param._type = decl._type
```

Comment:
This part is tricky. I think we could use some `assert(decl._type)`, for defensive programming.